botocore/handlers.py

```diff
@@ -269,13 +271,16 @@ def add_expect_header(model, params, **kwargs):
 def quote_source_header(params, **kwargs):
     if params['headers'] and 'x-amz-copy-source' in params['headers']:
         value = params['headers']['x-amz-copy-source']
-        p = urlsplit(value)
-        # We only want to quote the path.  If the user specified
-        # extra parts, say '?versionId=myversionid' then that part
-        # should not be quoted.
-        quoted = quote(p[2].encode('utf-8'), '/~')
-        final_source = urlunsplit((p[0], p[1], quoted, p[3], p[4]))
-        params['headers']['x-amz-copy-source'] = final_source
+        result = VERSION_ID_SUFFIX.search(value)
+        if result is None:
+            params['headers']['x-amz-copy-source'] = percent_encode(
+                value, safe=SAFE_CHARS + '/')
```

Comment:
It is looking like ``SAFE_CHAR`` is picking up some extra characters other than ``~``. Is this intentional?